plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -138,6 +149,23 @@ public void onRequestEnd(final SearchPhaseContext context, final SearchRequestCo
             attributes.put(Attribute.TOTAL_SHARDS, context.getNumShards());
             attributes.put(Attribute.INDICES, request.indices());
             attributes.put(Attribute.PHASE_LATENCY_MAP, searchRequestContext.phaseTookMap());
+
+            // Get internal computed and user provided labels
+            Map<String, Object> labels = new HashMap<>();
+            // Retrieve user provided label if exists
+            ThreadContext threadContext = threadPool.getThreadContext();
+            String userProvidedLabel = threadContext.getRequestHeadersOnly().get(Task.X_OPAQUE_ID);
+            if (userProvidedLabel != null) {
+                labels.put(Task.X_OPAQUE_ID, userProvidedLabel);
+            }
```

Comment:
The reading and writing for headers should be in one place

Reply:
Let me refactor the service to reflect this.